File: MaiBaiApplication/TianShenCash/src/main/java/com/tianshen/cash/activity/MyBankCardActivity.java
package com.tianshen.cash.activity;

import android.annotation.TargetApi;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.RelativeLayout;

import com.tianshen.cash.R;
import com.tianshen.cash.adapter.BankCardInfoAdapter;
import com.tianshen.cash.base.BaseActivity;
import com.tianshen.cash.constant.GlobalParams;
import com.tianshen.cash.model.GetBankListBean;
import com.tianshen.cash.net.api.GetBindBankList;
import com.tianshen.cash.net.base.BaseNetCallBack;
import com.tianshen.cash.net.base.UserUtil;
import com.tianshen.cash.utils.LogUtil;
import com.tianshen.cash.utils.TelephoneScreenChangeUtils;
import com.umeng.analytics.MobclickAgent;

import org.json.JSONException;
import org.json.JSONObject;

public class MyBankCardActivity extends BaseActivity implements View.OnClickListener {

    private ListView lv_bank_list;
    private Button bt_remove_binding;
    private RelativeLayout rl_add_bank_card;
    private Dialog dialog;
    private GetBankListBean bankListBean;
    private IntentFilter intentFilter;
    private BankRemoveBindReceiver bankRemoveBindReceiver;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setViewType(false);
        initBankInfo();
        intentFilter = new IntentFilter();
        intentFilter.addAction(GlobalParams.REMOVEBIND_SUCCESS_ACTION);
        intentFilter.addAction(GlobalParams.BIND_CARD_SUCCESS_ACTION);
        //创建NetWorkChangeReceiver的实例，并调用registerReceiver()方法进行注册
        bankRemoveBindReceiver = new BankRemoveBindReceiver();
        registerReceiver(bankRemoveBindReceiver, intentFilter);
    }

    @Override
    protected int setContentView() {
        return R.layout.activity_my_bank_card;
    }

    @Override
    protected void findViews() {
        lv_bank_list = (ListView) findViewById(R.id.lv_bank_list);
        bt_remove_binding = (Button) findViewById(R.id.bt_remove_binding);
        rl_add_bank_card = (RelativeLayout) findViewById(R.id.rl_add_bank_card);
    }

    @Override
    protected void setListensers() {
        bt_remove_binding.setOnClickListener(this);
        rl_add_bank_card.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.bt_remove_binding:
                //解除绑定
                showRemoveBindDialog();
                break;
            case R.id.rl_add_bank_card:
                //绑定银行卡
                bindBankCard();
                break;
        }
    }

    private void initBankInfo() {
        //执行获取银行卡列表的方法
        GetBindBankList getBindBankList = new GetBindBankList(mContext);
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("customer_id", UserUtil.getId(mContext));
            getBindBankList.getBindBankList(jsonObject, null, true, new BaseNetCallBack<GetBankListBean>() {
                @Override
                public void onSuccess(GetBankListBean paramT) {
                    bankListBean = paramT;
                    if (bankListBean.getData().size() > 0) {
                        setViewType(true);
                        BankCardInfoAdapter adapter = new BankCardInfoAdapter(mContext, bankListBean);
                        lv_bank_list.setAdapter(adapter);
                    } else {
                        setViewType(false);
                    }
                }

                @Override
                public void onFailure(String url, int errorType, int errorCode) {

                }
            });
        } catch (JSONException e) {
            e.printStackTrace();
            MobclickAgent.reportError(mContext, LogUtil.getException(e));
        }
    }

    private void setViewType(boolean isHadCard) {
        if (isHadCard) {
            lv_bank_list.setVisibility(View.VISIBLE);
            bt_remove_binding.setVisibility(View.VISIBLE);
            rl_add_bank_card.setVisibility(View.GONE);
        } else {
            lv_bank_list.setVisibility(View.GONE);
            bt_remove_binding.setVisibility(View.GONE);
            rl_add_bank_card.setVisibility(View.VISIBLE);
        }
    }

    private void bindBankCard() {
        //绑定银行卡
        Bundle bundle = new Bundle();
        bundle.putString(GlobalParams.REPAY_FROM_KEY, "bindBankCard");//用户主动绑定银行卡，不付款
        gotoActivity(mContext, AddBankCardActivity.class, bundle);
    }

    @TargetApi(Build.VERSION_CODES.KITKAT)
    private void showRemoveBindDialog() {
        dialog = new AlertDialog.Builder(mContext, R.style.remove_bind_dialog).create();
        View view = LayoutInflater.from(mContext).inflate(R.layout.view_dialog_remove_bind, null);
        LinearLayout ll_cancel = (LinearLayout) view.findViewById(R.id.ll_cancel);
        LinearLayout ll_confirm = (LinearLayout) view.findViewById(R.id.ll_confirm);
        ll_cancel.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                TelephoneScreenChangeUtils telephoneScreenChangeUtils = new TelephoneScreenChangeUtils(mContext);
                telephoneScreenChangeUtils.changeLight();
                dialog.dismiss();
            }
        });
        ll_confirm.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                TelephoneScreenChangeUtils telephoneScreenChangeUtils = new TelephoneScreenChangeUtils(mContext);
                telephoneScreenChangeUtils.changeLight();
                dialog.dismiss();
                if (bankListBean == null || bankListBean.getData().size() == 0) {
                    return;
                }
                Bundle mBundle = new Bundle();
                mBundle.putString("intent", "unBindCard");
                mBundle.putString("bank_card", bankListBean.getData().get(0).getCard_num());
                gotoActivity(mContext, InputPayPwdActivity.class, mBundle);
            }
        });
        dialog.setCancelable(false);
        TelephoneScreenChangeUtils telephoneScreenChangeUtils = new TelephoneScreenChangeUtils(mContext);
        telephoneScreenChangeUtils.changeDark();
        dialog.show();
        dialog.setContentView(view);
    }

    private class BankRemoveBindReceiver extends BroadcastReceiver {

        @Override
        public void onReceive(Context context, Intent intent) {
            switch (intent.getAction()) {
                case GlobalParams.BIND_CARD_SUCCESS_ACTION:
                case GlobalParams.REMOVEBIND_SUCCESS_ACTION:
                    bankListBean = null;
                    initBankInfo();
                    break;
            }
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unregisterReceiver(bankRemoveBindReceiver);
    }
}

File: MaiBaiApplication/TianShenCash/src/main/java/com/tianshen/cash/activity/MyBankCardActivity.kt
package com.tianshen.cash.activity

import android.support.v7.widget.LinearLayoutManager
import android.view.View
import com.afollestad.materialdialogs.MaterialDialog
import com.tianshen.cash.R
import com.tianshen.cash.adapter.MyBankCardAdapter
import com.tianshen.cash.base.BaseActivity
import com.tianshen.cash.constant.GlobalParams
import com.tianshen.cash.event.UserConfigChangedEvent
import com.tianshen.cash.model.GetBankListBean
import com.tianshen.cash.model.GetBankListItemBean
import com.tianshen.cash.model.PostDataBean
import com.tianshen.cash.model.UserAuthCenterBean
import com.tianshen.cash.net.api.GetBindBankList
import com.tianshen.cash.net.api.GetUserAuthCenter
import com.tianshen.cash.net.api.UnBindMyBankCard
import com.tianshen.cash.net.base.BaseNetCallBack
import com.tianshen.cash.utils.TianShenUserUtil
import com.tianshen.cash.utils.ToastUtil
import kotlinx.android.synthetic.main.activity_my_bank_card.*
import org.greenrobot.eventbus.EventBus
import org.json.JSONException
import org.json.JSONObject


class MyBankCardActivity : BaseActivity() {

    private var mAdapter: MyBankCardAdapter? = null
    private var identityStatus = 0

    private var isUnBind: Boolean = false

    override fun onResume() {
        super.onResume()
        initMyBankCardData()
        isUnBind = false
    }

    override fun setContentView() = R.layout.activity_my_bank_card

    override fun findViews() {
    }

    override fun setListensers() {
        //点击了返回
        tv_my_bank_card_back.setOnClickListener {
            backActivity()
        }

        //点击了添加银行卡
        ll_add_bank_card.setOnClickListener {
            if (0 == identityStatus) {
                ToastUtil.showToast(mContext, "请先进行身份认证")
            } else {
                gotoActivity(mContext, AuthBankCardActivity::class.java, null)
            }
        }

    }

    /**
     * 获取银行卡信息
     */
    private fun initMyBankCardData() {
        var getBindBankList = GetBindBankList(mContext)
        var jsonobject = JSONObject()
        var userId = TianShenUserUtil.getUserId(mContext)
        jsonobject.put(GlobalParams.USER_CUSTOMER_ID, userId)
        getBindBankList.getBindBankList(jsonobject, null, true, object : BaseNetCallBack<GetBankListBean> {
            override fun onSuccess(paramT: GetBankListBean?) {
                var size = paramT?.data?.size
                if (0 == size) {
                    if (isUnBind) {
                        EventBus.getDefault().post(UserConfigChangedEvent())
                    }
                    showAddBankCardUI()
                    initAuthCenterData()

                } else {
                    showMyBankCardUI(paramT)
                }
            }

            override fun onFailure(url: String?, errorType: Int, errorCode: Int) {

            }

        })

    }

    /**
     * 解绑银行卡
     */
    private fun unBind(cardNum: String) {
        var unBindMyBankCard = UnBindMyBankCard(mContext)
        var jsonobject = JSONObject()

        var userId = TianShenUserUtil.getUserId(mContext)
        jsonobject.put(GlobalParams.USER_CUSTOMER_ID, userId)
        jsonobject.put("card_num", cardNum)
        unBindMyBankCard.unBind(jsonobject, object : BaseNetCallBack<PostDataBean> {
            override fun onSuccess(paramT: PostDataBean?) {
                initMyBankCardData()
            }

            override fun onFailure(url: String?, errorType: Int, errorCode: Int) {

            }

        })
    }

    /**
     * 得到用户认证信息
     */
    private fun initAuthCenterData() {
        try {
            val jsonObject = JSONObject()
            val userId = TianShenUserUtil.getUserId(mContext)
            jsonObject.put(GlobalParams.USER_CUSTOMER_ID, userId)
            val getUserAuthCenter = GetUserAuthCenter(mContext)
            getUserAuthCenter.userAuthCenter(jsonObject, null, true, object : BaseNetCallBack<UserAuthCenterBean> {
                override fun onSuccess(paramT: UserAuthCenterBean) {
                    if ("0".equals(paramT.data.id_num) || "0".equals(paramT.data.face_pass)) {//判断身份认证和扫脸都成功没。如果有一个失败就算身份认证失败
                        identityStatus = 0
                    } else {
                        identityStatus = 1
                    }
                }

                override fun onFailure(url: String, errorType: Int, errorCode: Int) {}
            })

        } catch (e: JSONException) {
            e.printStackTrace()
        }

    }

    /**
     * 显示添加银行卡Ui
     */
    private fun showAddBankCardUI() {
        ll_add_bank_card.visibility = View.VISIBLE
        xrecyclerview_my_bank_card.visibility = View.GONE
    }

    /**
     * 显示我的银行卡UI
     */
    private fun showMyBankCardUI(paramT: GetBankListBean?) {

        ll_add_bank_card.visibility = View.GONE
        xrecyclerview_my_bank_card.visibility = View.VISIBLE

        if (mAdapter == null) {
            xrecyclerview_my_bank_card.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
            xrecyclerview_my_bank_card.setPullRefreshEnabled(false)
            xrecyclerview_my_bank_card.setLoadingMoreEnabled(false)
            mAdapter = MyBankCardAdapter(paramT?.data!!, {
                showUnBindBankCardDialog(it)
            })
            xrecyclerview_my_bank_card.adapter = mAdapter
        } else {
            mAdapter?.setData(paramT?.data!!)
            mAdapter?.notifyDataSetChanged()
        }
    }


    /**
     * 显示解绑dialog
     */
    private fun showUnBindBankCardDialog(bean: GetBankListItemBean) {

        var bankCardNum = bean.getCard_num()
        var bankCardNumLast = bankCardNum.substring(bankCardNum.length - 4, bankCardNum.length)
        var content = "您确定解除绑定尾号为 $bankCardNumLast 的银行卡吗"

        MaterialDialog.Builder(mContext)
                .content(content)
                .positiveText("确认")
                .negativeText("取消")
                .onPositive { _, _ ->
                    isUnBind = true
                    unBind(bankCardNum)
                }
                .onNegative { _, _ ->
                    isUnBind = false
                }
                .show()
    }

}
